stm32f3xx I2C, lsm303dlhc sensor and secondary drivers of ninedof

alexandruradovici opened the pull request:
### Pull Request Overview

This pull request:
  * adds I2C Master support for the stm32f3xx chips
  * adds the lsm303dlhc (3 axis accelerometer and magnetometer) capsule
  * adds the ability to use secondary drivers for the ninedof hil. Boards like the stm32f3discovery have all the three ninedof functionalities, but provided by two different hardware components. The modification allows the ninedof hil to use several hardware capsules to collect all the data.
  * adds the l3gd20 and lsm303dlhc sensors to the stm32f3discovery board
  * fixes the use imports issue for the l3gd20
  * fixes the imports for the i2c component 

### Testing Strategy

This pull request was tested using an stm32f3 discovery board

### TODO or Help Wanted

Any feedback is appreciated.


### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make formatall`.

bradjc commented on boards/stm32f3discovery/src/main.rs:
```diff
@@ -374,13 +436,80 @@ pub unsafe fn reset_handler() {
         led: led,
         button: button,
         alarm: alarm,
+        // uncomment this is you need low level acces to the l3gd20 sensor
```
```suggestion
        // uncomment this if you need low level access to the l3gd20 sensor
```

bradjc commented on boards/stm32f3discovery/src/main.rs:
```diff
@@ -374,13 +436,80 @@ pub unsafe fn reset_handler() {
         alarm: alarm,
+        // uncomment this is you need low level acces to the l3gd20 sensor
+        // l3gd20: l3gd20,
+        // uncomment this is you need low level acces to the lsm303dlhc sensor
```
```suggestion
        // uncomment this if you need low level access to the lsm303dlhc sensor
```

alexandruradovici commented:
> Overall looks good.
> 
>     * I think we should take a stand on the low level access to the sensors on the discovery platform, rather than make them optional through comments. Given it is a development platform, I think we should default to exposing the sensor drivers.
> 
Ok, will do.
>     * I like the evolution of the ninedof capsule and allowing "secondary" chips to complete the interface. What about just making all of the drivers part of a list, and dropping the notion of secondary? The semantics could be the same, but is there a need to special case the first sensor?
That was my initial intention, I thought about being backward compatible. If it is ok with you, I can change all the usage of ninedof and docs to work with the list of drivers.

bradjc commented on capsules/src/lsm303dlhc.rs:
```diff
@@ -0,0 +1,801 @@
+impl Driver for Lsm303dlhcI2C<'a> {
+    fn command(&self, command_num: usize, data1: usize, data2: usize, _: AppId) -> ReturnCode {
+        match command_num {
+            0 /* check if present */ => ReturnCode::SUCCESS,
```
```suggestion
            0 => ReturnCode::SUCCESS,
```
Helps with rustfmt issue.

alexandruradovici commented on doc/syscalls/70006_lsm303dlhc.md:
```diff
@@ -0,0 +1,129 @@
+
+    **Argument 2**: unused
+
+    **Returns**: `SUCCESS` if there is no other command in progress, `EBUSY` otherwise.
```
If present is async, it writes and reads from the device. The subscribed callback will return whether the device is present or not. The command just issues the writing and the reading. It can only do this if there is no other command in progress, thus the EBUSY.

ppannuto replied:
Ahh.. that makes sense -- my concern then I think is if someone is reading this it would be very easy to think that a return of `SUCCESS` means the hardware is present; I gave suggested updated language.

ppannuto replied:
```suggestion
    **Returns**: `SUCCESS` if presence test has started successfully, `EBUSY` otherwise.
```

alexandruradovici commented on doc/syscalls/70006_lsm303dlhc.md:
```diff
@@ -0,0 +1,129 @@
+
+    **Returns**: `SUCCESS` if there is no other command in progress, `EBUSY` otherwise.
+
+  * ### Command number: `2`
```
Yes, you have a point, I'll add the separate calls. Usually you want to configure the sensor upfront, but low power makes sense.

ppannuto replied:
@bradjc I think these changes are still missing?

alexandruradovici replied:
I would keep the API like this until we have modify_register API for I2C and update the driver after that.

alexandruradovici commented on doc/syscalls/70006_lsm303dlhc.md:
```diff
@@ -0,0 +1,129 @@
+
+    **Returns**: `SUCCESS` if there is no other command in progress, `EBUSY` otherwise.
+
+  * ### Command number: `4`
```
No, I didn't think of an order, I can change them.

alexandruradovici commented on chips/stm32f3xx/src/spi.rs:
```diff
@@ -271,13 +271,13 @@ impl Spi<'a> {
 
         if self.transfers.get() == SPI_IN_PROGRESS {
+            self.release_low();
+            self.transfers.set(SPI_IDLE);
```
I added a comment, it is exactly what you said.

alexandruradovici commented on doc/syscalls/70006_lsm303dlhc.md:
```diff
@@ -0,0 +1,129 @@
+
+    **Argument 1**: Accelerometer scale defined in manual table 27, page 27
+
+    **Argument 2**: High resolution (1 on, 0 off)
```
This might be useful, adding it.

alexandruradovici replied:
I'm not sure how to go about this. To enable low power, this means modifying a register, which implies reading and then writing back to it. Most of the options actually work like this. I am wondering whether to do this for several registers here, or maybe we could add this to the I2C HIL? I am guessing most of the I2C devices will have this pattern, so maybe implementing a functionality similar to the MCU registers might be useful.

Any thoughts here?

alexandruradovici replied:
I would keep the API like this until we have modify_register API for I2C and update the driver after that.

bradjc commented on boards/components/src/ninedof.rs:
```diff
@@ -0,0 +1,106 @@
+//! Components for NineDof
+//!
+//! Provides two components:
+//! 	1. NineDofComponent used ofr the ninedof driver
```
```suggestion
//! 	1. NineDofComponent used for the ninedof driver
```

bradjc commented on capsules/src/ninedof.rs:
```diff
@@ -51,21 +53,45 @@ impl Default for App {
+
 pub struct NineDof<'a> {
-    driver: &'a dyn hil::sensors::NineDof,
+    drivers: List<'a, NineDofNode<'a, &'a dyn hil::sensors::NineDof>>,
```
What about using something like LED as a template and using a slice instead? This would reduce complexity, and I don't think we need the flexibility of the List approach.

That would also simplify the component interface, and remove the possibility of forgetting to create a `NineDofDriverComponent`.

alexandruradovici replied:
Thank you for the feedback, make much more sens like this.

bradjc commented on chips/stm32f3xx/src/spi.rs:
```diff
@@ -271,13 +271,15 @@ impl Spi<'a> {
         }
 
         if self.transfers.get() == SPI_IN_PROGRESS {
+            // we elease the line and put the SPI in IDLE as the client might
```
```suggestion
            // we release the line and put the SPI in IDLE as the client might
```

ppannuto commented on doc/syscalls/70006_lsm303dlhc.md:
```diff
@@ -0,0 +1,130 @@
+
+  * ### Command number: `1`
+
+    **Description**: Verifies whether the hardware sensor is present.
```
```suggestion
    **Description**: Performs a test read/write to verify that the hardware sensor is present.
```

ppannuto commented on boards/stm32f3discovery/src/main.rs:
```diff
@@ -293,8 +350,8 @@ pub unsafe fn reset_handler() {
-        stm32f3xx::gpio::PinId::PA05.get_pin().as_ref().unwrap(),
-        stm32f3xx::gpio::PinId::PA07.get_pin().as_ref().unwrap(),
+        // stm32f3xx::gpio::PinId::PA05.get_pin().as_ref().unwrap(),
+        // stm32f3xx::gpio::PinId::PA07.get_pin().as_ref().unwrap(),
```
Should these just be deleted from here?

alexandruradovici replied:
Done

ppannuto commented on capsules/src/lsm303dlhc.rs:
```diff
@@ -0,0 +1,779 @@
+//! <https://www.st.com/en/mems-and-sensors/lsm303dlhc.html>
+//!
+//! Syscall Interface
+//! -----------------
```
I might suggest removing this here and putting a link to the documentation instead. I'm always wary when things are documented in two places, easy to accidentally get out of sync, and which is the authority?

alexandruradovici replied:
Are you suggesting to remove the syscall interface and place a link towards the markdown file?

ppannuto replied:
Yep!

alexandruradovici replied:
Done

ppannuto commented on capsules/src/lsm303dlhc.rs:
```diff
@@ -0,0 +1,779 @@
+/// Register values
+const REGISTER_AUTO_INCREMENT: u8 = 0x80;
+
+/// Accelerometer Registers
```
I think these might also benefit from using `enum_from_primitive`, then I think you wouldn't have to have the presently unused fellows commented out, and this `///` doc comment would document all of the enum members rather than this one constant

alexandruradovici replied:
Done

alexandruradovici replied:
Done.

alexandruradovici commented:
While adding the `set_low_power` and `read_low_power` functionalities as @ppannuto  suggested, I realized that several I2C devices need the _modify_register_ pattern, meaning that they will have to issue a read command, modify some bits and issue a write command. I tried to use the registers macros to define the I2C registers and added the following function and macros to the I2CDevice trait:

```rust
    /// Modify u8 register
    /// _buffer[0] is register address
    /// _buffer[1] is register address
    /// _buffer[2] is register new value
    /// _buffer[3] is register value mask
    /// use the i2c_modify_register macro to set the parameters
    fn modify_register(&self, _buffer: &'static mut [u8]) {
        panic!("i2c modify register is not implemented");
    }
```

I tried to add the `FieldValue` parameter, but as `FieldValue` is parametrized and the i2c hil is used as a trait object, this would not work. I realized that it is much easier to just provide a macro that receives the `FieldValue` and transforms it into the actual `u8` parameters required for `modify_register`.

```rust
#[macro_export]
macro_rules! i2c_modify_register {
    ($buffer: expr, $reg_address:expr, $field_value:expr) => {{
        $buffer[0] = $reg_address as u8;
        $buffer[1] = $reg_address as u8;
        $buffer[2] = $field_value.value;
        $buffer[3] = $field_value.mask;
        $buffer
    }};
}
```

This approach makes writing drivers much more expressive, as register fields are shown.

For reading back a value for a register field, I added another macro that receives the buffer returned by `command_complete` and returns the value of the field

```rust
#[macro_export]
macro_rules! i2c_read_register_field_value {
    ($buffer:expr, $field_value:expr) => {
        $field_value.val($buffer[0]).read($field_value)
    };

    ($buffer:expr, $index:expr, $field_value:expr) => {
        $field_value.val($buffer[$expr]).read($field_value)
    };
}
```

As an example, I used this to write the `set_low_power` and `read_low_power` functions.

I don't know whether this is the best approach, any feedback or ideas would very useful.

bradjc commented:
Can you edit your comment to include an example usage of the macro?

ppannuto commented on doc/syscalls/70006_lsm303dlhc.md:
```diff
@@ -0,0 +1,129 @@
+
+  * ### Subscribe number `0`
+
+    **Description**: Called when commands are done
```
<img width="528" alt="image" src="https://user-images.githubusercontent.com/339422/79894655-c71b1380-83d3-11ea-8067-24f67959afda.png">
nit: something is wonky with the whitespace here

alexandruradovici replied:
I think this is from my editor, is there any guideline for the md?

alexandruradovici replied:
By driver do you mean the i2c file from each chip? Wouldn't this require a change in all the chips, vs the current solution of implementing it in the virtual_i2c device?

alexandruradovici replied:
Any feedback here so we can we can merge all the functionalities for the discovery?

bradjc commented on capsules/src/virtual_i2c.rs:
```diff
@@ -90,6 +90,7 @@ pub struct I2CDevice<'a> {
     operation: Cell<Op>,
     next: ListLink<'a, I2CDevice<'a>>,
     client: OptionalCell<&'a dyn I2CClient>,
+    modify_register_in_progess: Cell<bool>,
```
What does this flag do?

alexandruradovici replied:
I separated the changes, for now the lsm303dlhc capsule is with a set only interface. As soon as we have a modify_register api figured out, I will add back the functionality.

alexandruradovici commented:
I separated the changes, for now the lsm303dlhc capsule is with a set only interface. As soon as we have a modify_register api figured out, I will add back the functionality.

alexandruradovici commented:
I think I addressed all of them. I suggested postponing the fiull API until we have a modify_register primitive for the I2C and deleted the syscall interface from the rs file and added a link to the md file. 

The link will be valid as soon as this gets merged.
